### ISSUE TITLE
ToDo list

### ISSUE THREAD

**Issue by jarun** [jarun/nnn#1022]
Rolled from #935.

### Cooking

- [x] a patch management model for approved patches
- [x] multi-threaded disk usage calculation using pthreads and FTS
- [x] dynamic view update when calculating disk usage
- [x] <kbd>Bksp</kbd>/<kbd>Del</kbd> at empty filter prompt to refresh dir
- [x] Try to create new context with <kbd>Shift-TAB</kbd>, else reverse cycle
- [x] <kbd>Alt+Esc</kbd> to quit context from filter prompt 
- [x] fix zombies left behind after running plugins (#999)
- [x] named persistent sessions (mcchrish/nnn.vim#43)
- [x] consider nanosecond resolution when sorting by time (#978)
- [x] check external selection in `'c'urrent / 's'el` prompt (#976)
- [x] show number of files selected in local selection buffer, if any
- [x] `nnn` & picker plugin (e.g. `fzopen`) sync (mcchrish/nnn.vim#82)
- [x] make batch rename interactive (#971)
- [x] cached uid/gid for performance improvement
- [x] fixes for `nnn` with `netbsd-curses`, `musl-fts`, `musl` (#998)
- [x] script to statically compile `nnn` with `musl gcc` on Ubuntu
- [x] restore hovered file when plugin is chosen from plugin dir
- [x] support QuickLook on WSL in `preview-tui` (#959)
- [x] toggle `preview-tui` with the same _custom_ plugin key
- [x] smoother preview toggling in `preview-tui` (#966)
- [x] `listen_on` should be set in kitty.conf for `preview-tui` (#970)
- [x] minimal `bat` style in `preview-tui`, honors `$BAT_STYLE`
- [x] plugin `preview-tui-ext` supersedes `preview-tui` (#1033)
- [x] plugin `fzcd` can now fuzzy search multiple directories
- [x] plugin `imgview` supersedes `imgthumb` and `vidthumb`
- [x] plugin `umounttree`: unmount remote mntpoint from within
- [x] plugin `xdgdefault`: set the default app for hovered file type
- [x] plugin `fzplug`: fuzzy find, preview and run other plugins
- [x] plugin `cmusq`: queue/play music in `cmus`
- [x] plugin `mocplay` renamed to `mocq`
- [x] plugin `cleanfilename` renamed to `fixname`
- [x] go to last dir on <kbd>~</kbd> (HOME) or <kbd>`</kbd> (ROOT) key repeat
- [x] ambiguous key <kbd>^Space</kbd> to select/clear range dropped (#998)
- [x] user wiki page for [Themes](https://github.com/jarun/nnn/wiki/Themes)
- [x] show selection mark (`+`) in reverse bold for improved visibility
- [x] reverse block replaces `>` to mark hovered entry in detail mode
- [x] make option `O_CKBOARD` removed
- [x] make option `O_NOLOC` renamed to `O_NOLC`
- [x] ignore `O_NOLC` if `O_ICONS` or `O_NERD` is specified (#1026)
- [x] unicode arrow indicators if `O_ICONS` or `O_NERD` is specified
- [x] make option `NOX11`: disable notis, sel-clipboard sync, xterm title
- [x] retain filter in _nav-to-type_ mode after file open
- [x] fix no files picked with `NNN_TMPFILE` exported and <kbd>q</kbd> to quit
- [x] disable xterm title setting in picker mode (#974)

### Up for grabs

- [ ] notcurses UI backend (#1014) [@KlzXS]
- [ ] a light colorscheme with `NNN_FCOLORS` (#937)
- [ ] update only current context data when saving session (#837)
- [ ] generate static binary for arm, aarch with releases using CI (#623)
- [ ] support applying filter by key (like bookmarks/plugins)
- [ ] open new terminal window in spawned shell directory (#759)

For anything else please discuss in this thread.

**Contribution [guideline](https://github.com/jarun/nnn/blob/master/misc/CONTRIBUTING.md).**

**Comment by mizlan** [comment]
> predefined filters like bookmarks

I've read a past thread but am still confused as to what this means.

**Comment by jarun** [comment]
Updated now. Basically filter keys to apply a filter at the start of the program. However, I don't see much benefit.
See if the multi-threaded `du` interests you.

However, I would like the `nnn.vim` plugin enhancement to support sessions first.

**Comment by jarun** [comment]
@Kabouik can you please audit `nuke` in the lines of https://github.com/jarun/nnn/issues/935#issuecomment-841088727?

It would be good to have that done before the next release.

**Comment by jarun** [comment]
Leave it. I took a look, it's more or less clean.

**Comment by jayp0501** [comment]
I need someone to help me to try to build and go through this I have issues with readingBut I do have someone that is going through my phone blocking me from things and I need to put a stop to it they help me where I cannot advertise on craigslist they had me where I was not able to go on some websites and I will stop at Jewett

**Comment by jarun** [comment]
@CodeforEvolution does Haiku have libpthread? Planning to implement threaded disk usage calculator.

**Comment by jarun** [comment]
Note that we will also need fts.

**Comment by CodeforEvolution** [comment]
Sure thing! And yes, it looks like we have fts. :)

**Comment by jarun** [comment]
Cool! This is very important for the next release.

**Comment by CodeforEvolution** [comment]
@jarun Well, better yet, I can confirm that no additional changes are required! I was able to successfully build the latest commit of nnn and run the test you posted in the commit message.

**Comment by jarun** [comment]
Just awesome!!! :+1:

**Comment by luukvbaal** [comment]
Not sure what .thumbs is but no, thumbnails go to `$TMPDIR` by default and a suggestion to set `$NNN_PREVIEWDIR` to `$XDG_CACHE_HOME/nnn/previews` is present in the docs.

**Comment by Stig124** [comment]
Hello,
I can make a multi-arch docker container to build the arm images if needed

**Comment by jarun** [comment]
@luukvbaal 

> Not sure what .thumbs is but no.

Thanks for confirming! Yes, I confirmed it's not from our previewers.

> I can make a multi-arch docker container to build the arm images if needed

@KlzXS @luukvbaal @0xACE what do you guys say?

**Comment by 0xACE** [comment]
> @Stig124
> 
> > I can make a multi-arch docker container to build the arm images if needed
> 
> @KlzXS @luukvbaal @0xACE what do you guys say?

Not sure what to say about it. Do as you please...

**Comment by KlzXS** [comment]
If it can be nicely integrated into your release workflow and produces a useable build, sure.

Do we have someone with these architectures? All I have that is not x86 is a pi zero.

**Comment by jarun** [comment]
I have a Pi 4 but can't guarantee I'll be able to test on every release.
However, I think it's a good idea in general if we can auto-generate these. We do have a line item.

I think @Stig124 you cna go ahead. Please do the CI integration as well.

**Comment by Stig124** [comment]
Will do

**Comment by KlzXS** [comment]
> can't guarantee I'll be able to test on every release.

Users will test it out in time. Let's just make sure the first binary we release doesn't have any major issues.

**Comment by jarun** [comment]
The next release is going to be a massive one with the features that are in.

**Comment by 0xACE** [comment]
One of my `ARM Cortex-A73` is running on my personal branch based on 950a8f6a6560f32a5f8737922e03c7bc0bdc9bd5 and it has been running fine for 3 years, nor have i experienced any issues exclusive to my `arm` machines...

**Comment by Stig124** [comment]
Had time to tinker with the image, it builds just fine, but I'm afraid that it won't go fine in CircleCI

**Comment by Stig124** [comment]
Hello,

I thank about that, it will not be possible to build executables on ARM using CircleCI, I spoke of that w/my boss, and he can loan me for free a ARM64-based machine that I can configure to respond to webhook so that It will be building the executables for nnn in both ARM64 and ARM32, what do you think @jarun?

**Comment by KlzXS** [comment]
The main issue with that becomes whether that machine will be available in the future or not. It would really suck if suddenly the arm builds stopped being built.

**Comment by jarun** [comment]
Can we use some service other than CircleCi? How about GitHub's own service?

**Comment by Stig124** [comment]
> Can we use some service other than CircleCi? How about GitHub's own service?

Travis seems to build ARM executable just fine, but I don't see any that would build armhf executables

**Comment by jarun** [comment]
In that case, we can drop the plan. With the very few deps `nnn` has, compiling it locally is not an issue.

**Comment by Stig124** [comment]
> In that case, we can drop the plan. With the very few deps `nnn` has, compiling it locally is not an issue.

I think this would be the best thing to do

**Comment by jarun** [comment]
@Stig124 does this help: https://circleci.com/blog/managing-ci-cd-pipelines-with-arm-compute-resource-classes/

**Comment by jarun** [comment]
Also see https://github.com/falcosecurity/falco/issues/1445

**Comment by Kabouik** [comment]
> Not sure what `.thumbs` is but no. Thumbnails go to `$TMPDIR/nnn/previews` by default and a suggestion to set `$NNN_PREVIEWDIR` to `$XDG_CACHE_HOME/nnn/previews` is present in the docs.

When I follow the suggestion to set `$NNN_PREVIEWDIR` either to the XDG variable or hardcoded to `~/.cache`, the preview directory gets created in the current directory for me, not in `~/.cache` as I would expect. Is that the expected behaviour? Could be a local env issue too, unfortunately I was on the move and haven't had the time to investigate more.

Other things I noticed, the `fzplug` detection for `bat` is reversed and uses `cat` when `bat` is detected. I can submit a PR for that soon. `preview-tui-*` also uses `cat` on my machine despite `bat` being installed, but there I couldn't fix it with just reversing the condition. Do you observe the same?

**Comment by luukvbaal** [comment]
I don't observe any of the thins you mention no. Are you sure you're on master? How are you settings those environment variables?

**Comment by luukvbaal** [comment]
It would place the previews in the current directory is empty if your `NNN_PREVIEWDIR` is somehow empty. Please let me know how you set your environment variable so I can try to reproduce. I set them in my `.zprofile` and never experienced any of those issues.

The master version of `fzplug` contains
```
# Preview with bat if installed
if type bat >/dev/null; then
    BAT="bat --terminal-width='$(tput cols)' --decorations=always --color=always --style='${BAT_STYLE:-header,numbers}'"
fi
...
--preview "${BAT:-cat}
```
Not sure how that could ever result in reversed detection. Same for `preview-tui` with
```
        if [ "$cmd" = "pager" ]; then
            if exists bat; then
                bat --terminal-width="$(tput cols)" --decorations=always --color=always \
                    --paging=never --style="$BAT_STYLE" "$@" &
```
Don't see how you could experience the behavior you describe if you're on master.

**Comment by Kabouik** [comment]
Yes, I'm on `master` with just a few changes on `fzplug` pending for a future PR, but none of them has to do with `bat` detection:

```diff
iff --git a/plugins/fzplug b/plugins/fzplug
index 11dcf7f..2e7f4c8 100755
--- a/plugins/fzplug
+++ b/plugins/fzplug
@@ -10,7 +10,7 @@
 #
 # Note: For better compatibility with as many nnn plugins as possible,
 #       fzplug will first execute the chosen script on the file hovered
-#       in nnn, and upon failure, try to run it with no target (i.e on
+#       in nnn, and upon failure, try to run it with no target (i.e., on
 #       an active selection, if present).
 #
 # Shell: POSIX compliant
@@ -31,16 +31,16 @@ nnnpluginsdir="$HOME/.config/nnn/plugins"
 
 # Preview with bat if installed
 if type bat >/dev/null; then
-    BAT="bat --terminal-width='$(tput cols)' --decorations=always --color=always --style='${BAT_STYLE:-header,numbers}'"
+    BAT="bat --terminal-width='$(tput cols)' --decorations=always --color=always --style='${BAT_STYLE:-header}'"
 fi
 
 plugin=$(find "$nnnpluginsdir" "$CUSTOMDIR1" "$CUSTOMDIR2" \
 -maxdepth 3 -perm -111 -type f 2>/dev/null | fzf --ansi --preview \
     "${BAT:-cat} {}" --preview-window="right:66%:wrap" --delimiter / \
-    --with-nth -1 --bind="?:toggle-preview")
+    --with-nth -1 --bind="?:toggle-preview,ctrl-e:execute(${EDITOR:-vi} {})")
 
 # Try running the script on the hovered file, and abort
-# abort if no plugin was selected (ESC or ^C pressed).
+# if no plugin was selected (ESC or ^C pressed).
 err=0
 if ! [ "$plugin" = "" ]; then
     "$plugin" "$1" || err=1
@@ -55,5 +55,5 @@ fi
 # Abort and show error if both fail
 if [ "$err" -eq "2" ]; then
     sep="\n---\n"
-    printf "$sep""Failed to execute '%s'. See error above or try without fzfplug. Press return to continue. " "$plugin" && read -r _ && clear
+    printf "$sep""Failed to execute '%s'. See error above or try without fzplug. Press return to continue. " "$plugin" && read -r _ && clear
 fi
```

If I replace `type bat` with `! type bat`, then it works correctly. However I'm not sure why `bat` doesn't work with `preview-tui*` either. Note that I have no `BAT_STYLE` set in my `.bashrc`, this might be a difference with your setup.

Regarding `NNN_PREVIEWDIR`, it was set this way in my `.bashrc`:

```bash
export NNN_PREVIEWDIR='~/.cache/nnn/previews'
```

But when I use that, I can see the folder be created in my current directories instead of `~/.cache`.

> It would place the previews in the current directory if your NNN_PREVIEWDIR is somehow empty. 

Hum, I didn't manually create `nnn/previews` after setting the environment variable and using previews in nnn, so maybe that's the reason.

**Comment by luukvbaal** [comment]
`NNN_PREVIEWDIR` being empty meaning the environment variable as seen by `preview-tui`. That must be what's going on if it's creating the previews in `PWD` i.e. `mkdir -p "$NNN_PREVIEWDIR/${3%/*}"`. No idea why that would be the case if you're on master and exporting `NNN_PREVIEWDIR` in your `.bashrc`, and `bash` is your default shell.

**Comment by Kabouik** [comment]
`type bat` returns `bat is /home/user/.cargo/bin/bat` on my end, not `0`. `! type bat` too, but that inverts the condition I guess.

**Comment by luukvbaal** [comment]
I'm talking about the exit status, we're redirecting the actual return value to `>/dev/null`.

**Comment by luukvbaal** [comment]
It must not return `/home/user/.cargo/bin/bat` in your "`nnn` plugin environment" or else you wouldn't have to invert the condition.

**Comment by Kabouik** [comment]
I think the issue with `NNN_PREVIEWDIR` is simply that it doesn't accept the `~/` abbreviation. I got confused by my other nnn environment variables where `~/` works (like plugins, bookmarks).

One issue though is that if you preview images in the preview directory, it will create another subdirectory for that, but then that means it can loop on itself. Wouldn't it be best to prevent `preview-tui*` from creating previews cache for images viewed inside a sub-folder of `$NNN_PREVIEWDIR`? I know no one should do that, but I feel it would be safer to keep the script from doing that.

**Comment by luukvbaal** [comment]
I've come across that situation while developing the plugin yeah but indeed no one should open `preview-tui` in the cache directory. Not sure if it's worth preventing that, perhaps it is.

**Comment by luukvbaal** [comment]
Indeed always prefer using `$HOME` over `~` in shell variables. `~` is not expanded when quoted.

**Comment by Kabouik** [comment]
I feel it could be useful even if it's weird to preview that folder on purpose. If the preview pane is already opened, then anything hovered will be previewed, meaning one would have to know about that issue in advance before going into the cache folder with nnn to remember that the preview pane should be closed beforehand. Going into that folder is not a rare use case I think, I would expect people to keep an eye on it to monitor its size and delete unnecessary cache.

**Comment by Kabouik** [comment]
> It must not return `/home/user/.cargo/bin/bat` in your "`nnn` plugin environment" or else you wouldn't have to invert the condition.

Right, I was just testing the command directly in terminal here, I misunderstood from your message that `0` was the expected result. The above reports in my previous messages are all from executing the script through `nnn` though, and adding `!` does fix the use of `bat` on that machine. Can't tell yet if I observe the same on other machines, but I assume I won't since you have no such issue on yours.

**Comment by Kabouik** [comment]
> Indeed always prefer using `$HOME` over `~` in shell variables. `~` is not expanded when quoted.

Thanks! Yeah I already use `$HOME` everywhere else in my `.bashrc` but got spoiled for nnn variables because nnn tends to be flexible: https://github.com/jarun/nnn/wiki/Basic-use-cases#add-bookmarks

Maybe we should edit the wiki to discourage this use, or is there a way to make `NNN_PREVIEWDIR` equally tolerant to `~/`?

**Comment by luukvbaal** [comment]
Yeah `export NNN_BMS="d:$HOME/Documents;u:$HOME/Cam Uploads;D:$HOME/Downloads/"` is more appropriate as an environment variable in my opinion. Not sure how @jarun feels about it.

**Comment by jarun** [comment]
Can you summarize the NNN_BMS thing for me?

**Comment by luukvbaal** [comment]
@Kabouik ran into an issue with `preview-tui` because they used "~" in `NNN_PREVIEWDIR` like suggested in the [wiki](https://github.com/jarun/nnn/wiki/Basic-use-cases#add-bookmarks) for `NNN_BMS`. This does not work for shell scripts without manually expanding "~", using "$HOME" instead is more appropriate.

**Comment by jarun** [comment]
Why can't the script expand `~`?

**Comment by luukvbaal** [comment]
Shells don't expand quoted `~`'s.

**Comment by luukvbaal** [comment]
See e.g. https://stackoverflow.com/questions/40223060/home-vs-for-use-in-bash-scripts

**Comment by luukvbaal** [comment]
Nothing needs to change, it's just a matter of whether you want to change the wiki to reflect using `$HOME` over `~` to possibly avoid users running into the same problem.

**Comment by jarun** [comment]
OK. I was expecting at least double quotes to work. Why can't you specify the unquoted string in a controlled script?

If space is a problem, doesn't the following work?

~/"Arun Pictures"/something

Anyway, the `NNN_BMS` thing stays.

**Comment by jarun** [comment]
> it's just a matter of whether you want to change the wiki to reflect using $HOME over ~ to possibly avoid users running into the same problem.

@Kabouik mentioned `NNN_BMS` which is parsed by `nnn`. It resolves `~` but doesn't understand `$HOME`. hence the explicit `~` usage in the example. I think @Kabouik is mixing up stuff. The shell limitation is not `nnn` issue.

**Comment by luukvbaal** [comment]
`$HOME` doesn't need to be parsed by `nnn` as it is expanded by the shell when assigned to a variable.

**Comment by luukvbaal** [comment]
It's just generally more advised to use `$HOME` in shell variables as it works in all cases as opposed to `~`.

**Comment by luukvbaal** [comment]
@Kabouik is indeed at fault for using `~` in `NNN_PREVIEWDIR` although I can understand why he would expect it to work when it does work for `NNN_BMS` since it is parsed by `nnn`.

Again, changing the wiki to use `$HOME` instead of `~` could prevent other users making the same mistake.

**Comment by jarun** [comment]
OK. Please update. You may have t do it in the man page as well (for `NNN_BMS`).

**Comment by Kabouik** [comment]
> I think @Kabouik is mixing up stuff. The shell limitation is not `nnn` issue.

No that's my understanding too now, but yeah initially I was thinking I could use the same syntax in environment variables parsed directly by `nnn`  or by its plugins, since all those environment variables look similar (`NNN_*`), which led me to misuse `NNN_PREVIEWDIR`.

I just find it misleading that the wiki shows `~` abbreviation working for variables parsed directly by `nnn` whereas it won't for those parsed by scripts, but I don't think every `nnn` user would be as ignorant as me when it comes to shells. :> I have no strong opinion about whether the wiki should show `$HOME` to avoid the ambiguity, or if scripts could be made to work with `~`  in environment variables, I just feel harmonizing would save mistakes for some users.

**Comment by jarun** [comment]
That's fine. It's good to be uniform.
